worker/changestream/metrics.go

```diff
@@ -0,0 +1,150 @@
+	ChangesRequestDuration *prometheus.HistogramVec
+	ChangesCount           *prometheus.HistogramVec
+	// EventMultiplexer metrics.
+	Subscriptions    prometheus.Gauge
+	DispatchDuration *prometheus.HistogramVec
+	DispatchErrors   *prometheus.CounterVec
+}
+
+func (c *Collector) WatermarkInsertsInc() {
+	c.WatermarkInserts.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) WatermarkRetriesInc() {
+	c.WatermarkRetries.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) ChangesRequestDurationObserve(val float64) {
+	c.ChangesRequestDuration.WithLabelValues(c.Namespace).Observe(val)
+}
+
+func (c *Collector) ChangesCountObserve(val int) {
+	c.ChangesCount.WithLabelValues(c.Namespace).Observe(float64(val))
+}
+
+func (c *Collector) SubscriptionsInc() {
+	c.Subscriptions.Inc()
+}
+
+func (c *Collector) SubscriptionsDec() {
+	c.Subscriptions.Dec()
+}
+
+func (c *Collector) SubscriptionsClear() {
+	c.Subscriptions.Set(0)
+}
+
+func (c *Collector) DispatchDurationObserve(val float64) {
+	c.DispatchDuration.WithLabelValues(c.Namespace).Observe(val)
+}
+
+func (c *Collector) DispatchErrorsInc() {
+	c.DispatchErrors.WithLabelValues(c.Namespace).Inc()
+}
+
+func (c *Collector) ForNamespace(namespace string) *Collector {
+	child := *c
+	child.Namespace = namespace
+
+	return &child
+}
+
+// NewMetricsCollector returns a new Collector.
+func NewMetricsCollector() *Collector {
+	return &Collector{
+		// Stream metrics.
+		WatermarkInserts: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: changestreamMetricsNamespace,
+			Subsystem: changestreamSubsystemNamespace,
+			Name:      "watermark_inserts_total",
+			Help:      "Total number of watermark insertions on the changelog witness table.",
+		}, labelNames),
+		WatermarkRetries: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: changestreamMetricsNamespace,
+			Subsystem: changestreamSubsystemNamespace,
+			Name:      "watermark_retries_total",
+			Help:      "Total number of watermark retries on the changelog witness table.",
+		}, labelNames),
+		ChangesRequestDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: changestreamMetricsNamespace,
+			Subsystem: changestreamSubsystemNamespace,
+			Name:      "changestream_duration_seconds",
+			Help:      "Total time spent in changestream requests.",
+		}, labelNames),
+		ChangesCount: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: changestreamMetricsNamespace,
+			Subsystem: changestreamSubsystemNamespace,
+			Name:      "changestream_count",
+			Help:      "Total number of changes returned by the changestream requests.",
+		}, labelNames),
+		// EventMultiplexer metrics.
+		Subscriptions: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace: changestreamMetricsNamespace,
+				Subsystem: changestreamSubsystemNamespace,
+				Name:      "subscription_count",
+				Help:      "The total number of subscriptions.",
+			},
+		),
```

Comment:
This should be a GuageVec with the namespace as well. As it's per model.

Reply:
nice catch!!! done